fdbserver/TagPartitionedLogSystem.actor.cpp

```diff
@@ -2041,40 +2047,128 @@ ACTOR Future<Void> TagPartitionedLogSystem::getDurableVersionChanged(LogLockInfo
+		int tLogIdx = 0;
+		boost::dynamic_bitset<> availableTLogs(bsSize);
+		// version -> tLogs (that are avaiable) that have received the version
+		std::unordered_map<Version, boost::dynamic_bitset<>> versionAvailableTLogs;
+		// version -> all tLogs that the version was sent to (by the commit proxy)
+		std::map<Version, boost::dynamic_bitset<>> versionAllTLogs;
+		// version -> prevVersion (that was given out by the sequencer) map
+		std::map<Version, Version> prevVersionMap;
 		int replicationFactor = std::get<0>(logGroupResult);
 		for (auto& tLogResult : std::get<1>(logGroupResult)) {
+			int tLogLocId = tLogLocIds[tLogGroupIdx][tLogIdx];
+			availableTLogs.set(tLogLocId);
 			bool logGroupCandidate = false;
 			for (auto& unknownCommittedVersion : tLogResult.unknownCommittedVersions) {
 				Version k = std::get<0>(unknownCommittedVersion);
 				if (k > minEnd) {
-					versionRepCount[k]++;
-					versionTLogCount[k] = std::get<1>(unknownCommittedVersion);
+					if (versionAvailableTLogs[k].empty()) {
+						versionAvailableTLogs[k].resize(bsSize);
+					}
+					versionAvailableTLogs[k].set(tLogLocId);
+					prevVersionMap[k] = std::get<1>(unknownCommittedVersion);
+					if (versionAllTLogs[k].empty()) {
+						versionAllTLogs[k].resize(bsSize);
+					}
+					populateBitset(versionAllTLogs[k], std::get<2>(unknownCommittedVersion));
 					logGroupCandidate = true;
 				}
 			}
 			if (!logGroupCandidate) {
 				return minEnd;
 			}
+			tLogIdx++;
 		}
 		Version minTLogs = minEnd;
-		for (auto const& [version, tLogCount] : versionTLogCount) {
-			if (versionRepCount[version] >= tLogCount - replicationFactor + 1) {
-				minTLogs = version;
+		Version prevVersion = minEnd;
+		for (auto const& [version, tLogs] : versionAllTLogs) {
+			if (prevVersion == minEnd || prevVersion == prevVersionMap[version]) {
+				// This version is not recoverable if there is a log server (LS) such that:
+				// - the commit proxy sent this version to LS (i.e., LS is present in "versionAllTLogs[version]")
+				// - LS is available (i.e., LS is present in "availableTLogs")
+				// - LS didn't receive this version (i.e., LS is not present in "versionAvailableTLogs[version]")
+				if (((tLogs & availableTLogs) & ~versionAvailableTLogs[version]).none()) {
+					// If the commit proxy sent this version to "N" log servers then at least
+					// (N - replicationFactor + 1) log servers must be available.
+					if (versionAvailableTLogs[version].size() >= tLogs.size() - replicationFactor + 1) {
+						minTLogs = version;
+						prevVersion = version;
+						// Update RVs.
+						for (boost::dynamic_bitset<>::size_type id = 0; id < versionAvailableTLogs[version].size();
+						     id++) {
+							if (versionAvailableTLogs[version][id]) {
+								RVs[id] = version;
+							}
+						}
+					} else {
+						break;
+					}
+				} else {
+					break;
+				}
 			} else {
 				break;
 			}
 		}
 		minLogGroup = std::min(minLogGroup, minTLogs);
```

Comment:
Same question here too: we are including the recovery version of the primary satellite in the cluster recovery version computation, not sure why we are doing it this way?